File: haskell/src/Xor.hs
--------------------------------------------------------------------------------
-- | XOR and XOR-decryption operations.
--
-- NOTE: This module was written with ASCII characters and strings in mind.
--       The ASCII assumption allows for quick resolutions when decrypting.
--       UTF8 is possible but more costly, so I'm not bothering with it here.
--
--------------------------------------------------------------------------------

module Xor
  ( -- * XOR operations
    xorEqLen, xorCycle
    -- * XOR decryption
  , xorDecrypt, xorDecryptWithKeys
  ) where

import LibCommon
import Bytes
import Hamming

import Data.Array.Unboxed
import Data.Char
import Data.List
import Data.List.Split
import Data.Maybe
import Data.Ord
import Data.Ratio

type Stats = Array Byte Int

-- | XOR two `ByteString`s of equal lengths.
--
-- Produces an error if the lengths differ.
xorEqLen :: ByteString -> ByteString -> ByteString
xorEqLen (a:as) (b:bs) = a .^. b : xorEqLen as bs
xorEqLen []     []     = []
xorEqLen _      _      = error "xorEqLen: ByteStrings have different lengths"

-- | XOR a `ByteString` (typically a text or cipher) with another `ByteString`,
-- repeating (cycling) the second one (typically a key).
xorCycle :: ByteString -> ByteString -> ByteString
xorCycle cipher = zipWith (.^.) cipher . cycle

-- | Tries to decrypt a repeating-key XOR cipher, given a set of candidate keys
-- and an ordered `String` of the most frequent ASCII characters expected.
--
-- Returns the key with the highest grade, as well as the grade (higher is better)
-- and the resulting plaintext, or `Nothing` if no key met the minimum criteria.
xorDecryptWithKeys :: [ByteString]  -- ^ Set of candidate keys.
                   -> String        -- ^ Ordered `String` of the most frequent ASCII characters expected.
                   -> ByteString    -- ^ Ciphertext.
                   -> Maybe (ByteString, Int, ByteString) -- ^ Key, grade and plaintext.
xorDecryptWithKeys keys topChars cipher = listToMaybe . sortBy (flip $ comparing snd3) $ solutions
  where topLen    = length topChars
        topBytes  = fromString topChars
        solutions = [(key, gradeKey, plaintext) | key <- keys
                    , let plaintext = cipher `xorCycle` key
                    , let stats     = asciiStats plaintext
                    , let histogram = take topLen . genHistogram $ fromJust stats
                    , let gradeKey  = grade histogram
                    , isJust stats, hasMatches topBytes histogram]
          where genHistogram stats  = sortBy (flip (comparing snd)) [(b, stats ! b) | b <- [0..byteMax]]
                hasMatches byteList = (==) minLen . length . take minLen . filter ((`elem` byteList) . fst)
                  where minLen = 2
                grade = foldl' (\acc (i, (b, _)) -> acc + gradeByte i b) 0 . zip [0..]
                  where gradeByte i b = case b `elemIndex` topBytes of
                          Just j  -> topLen - abs (i - j)
                          Nothing -> -topLen

-- | Creates a table of the number of occurences of the acceptable ASCII characters.
--
-- Returns `Nothing` upon encountering a non-ASCII or unusual character.
asciiStats :: ByteString -> Maybe Stats
asciiStats = go $ array (0,byteMax) [(i, 0) | i <- [0..byteMax]]
  where go stats []        = Just stats
        go stats (b:bs)
          | isBadAscii ! b = Nothing
          | otherwise      = go stats' bs
          where stats' = stats // [(b, 1 + stats ! b)]

-- | Highest accepted ASCII value.
asciiMax :: Int
asciiMax = ord '~'

-- | Highest accepted byte value.
byteMax :: Byte
byteMax = fromIntegral asciiMax

-- | Truth table of non-acceptable `Byte`s.
isBadAscii :: Array Byte Bool
isBadAscii = listArray (0, 255) [isBadAscii' b | b <- [0..255]]
  where isBadAscii' b = b > asciiMax
          || (b < ord ' ' && (b /= ord '\t' && b /= ord '\n' && b /= ord '\r' && b /= ord '\ESC'))

-- | Tries to decrypt a repeating-key XOR cipher, given an ordered `String` of
-- the most frequent ASCII characters expected.
--
-- Returns the key if all the blocks were decrypted successfully.
-- Returns `Nothing` otherwise.
xorDecrypt :: String -> ByteString -> Maybe ByteString
xorDecrypt topChars cipher = mapAllJust (head . fst3) solutions
  where keySize   = keysize cipher
        keys      = [[b] | b <- [0..255::Byte]]
        blocks    = transpose . chunksOf keySize $ cipher
        solutions = map (xorDecryptWithKeys keys topChars) blocks

-- | Finds the most likely keysize for a repeating-key XOR cipher.
-- Selects the keysize with the smallest normalized mean hamming distance
-- over the cipher.
keysize :: ByteString -> Int
keysize cipher = head . sortOn weightedMeanHammingDistance $ [2..maxKeySize]
  where cipherLen  = length cipher
        maxKeySize = min 40 (cipherLen `quot` 2)
        weightedMeanHammingDistance keySize = htot % (keySize * (nFullChunks - 1))
          where htot = sum . map (uncurry (^+)) . slidingPairs . take nFullChunks . chunksOf keySize $ cipher
                nFullChunks = cipherLen `quot` keySize

File: rust/src/xor.rs
use std::error::Error;
use std::fmt::{Display, Formatter, Result as FmtResult};

use self::XORError::*;

pub trait XOR<T: ?Sized> {
    fn xor(&mut self, other: &T) -> Result<(), XORError>;
    fn xor_cycle(&mut self, other: &T);
}

impl<T, O> XOR<O> for T
    where T: AsMut<[u8]>,
          O: AsRef<[u8]> {
    fn xor(self: &mut T, other: &O) -> Result<(), XORError> {
        if self.as_mut().len() == other.as_ref().len() {
            for (b1, &b2) in self.as_mut().iter_mut().zip(other.as_ref().iter()) {
                *b1 ^= b2;
            }
            Ok(())
        }
        else { Err(BadLength) }
    }

    fn xor_cycle(&mut self, other: &O) {
        for (b1, &b2) in self.as_mut().iter_mut().zip(other.as_ref().iter().cycle()) {
            *b1 ^= b2;
        }
    }
}

#[derive(Debug)]
pub enum XORError {
    BadLength,
}

impl Display for XORError {
    fn fmt(&self, f: &mut Formatter) -> FmtResult {
        write!(f, "XORing error: {}", self.description())
    }
}

impl Error for XORError {
    fn description(&self) -> &str {
        match *self {
            BadLength => "data length differ",
        }
    }
}

pub trait XORDecrypt<T>: XOR<T> {
    fn xor_decrypt_with_keys(&self, keys: &[T], frequent: &[u8]) -> Option<(usize, i32)>;
}

impl<T, K> XORDecrypt<K> for T
    where T: AsRef<[u8]> + AsMut<[u8]>,
          K: AsRef<[u8]> {
    fn xor_decrypt_with_keys(&self, keys: &[K], frequent: &[u8]) -> Option<(usize, i32)> {
        let mut solutions = Vec::new();
        'key: for (i, key) in keys.iter().enumerate() {
            let mut cipher = self.as_ref().to_owned();
            let mut stats = vec![0; 256];
            cipher.xor_cycle(key);
            for &b in &cipher {
                if bad_ascii(b as char) { continue 'key }
                stats[b as usize] += 1;
            }
            let mut histogram = (0..256).map(|b| (b as u8, stats[b])).collect::<Vec<_>>();
            histogram.sort_by(|&(_, ref count1), &(_, count2)| count2.cmp(count1));
            histogram.truncate(frequent.len());
            if 2 <= histogram.iter().filter(|&&(ref b, _)| frequent.contains(b)).count() {
                let grade = grade(&histogram, frequent);
                solutions.push((i, grade));
            }
        }
        solutions.sort_by(|&(_, ref grade1), &(_, grade2)| grade2.cmp(grade1));
        if solutions.len() != 0 { Some(solutions[0]) }
        else { None }
    }
}

fn bad_ascii(b: char) -> bool {
    b > '~' || (b < ' ' && (b != '\t' && b != '\n' && b != '\r' && b != 0x1b as char))
}

fn grade(histogram: &[(u8, usize)], frequent: &[u8]) -> i32 {
    let mut grade = 0;
    for (i, &(b, _)) in histogram.iter().enumerate() {
        if let Some(j) = index(b, frequent) {
            grade += frequent.len() as i32 - (i as i32 - j as i32).abs();
        }
        else { grade -= frequent.len() as i32; }
    }
    grade
}

fn index(b: u8, frequent: &[u8]) -> Option<usize> {
    frequent.iter().enumerate().filter_map(|(i, &c)| if b == c { Some(i) } else { None }).next()
}
